server/src/main/java/org/elasticsearch/transport/LeakTracker.java

```diff
@@ -0,0 +1,273 @@
+        @SuppressWarnings("unused")
+        private volatile int droppedRecords;
+
+        private final Set<Leak<?>> allLeaks;
+        private final int trackedHash;
+
+        private Leak(Object referent, ReferenceQueue<Object> refQueue, Set<Leak<?>> allLeaks) {
+            super(referent, refQueue);
+
+            assert referent != null;
+
+            trackedHash = System.identityHashCode(referent);
+            allLeaks.add(this);
+            headUpdater.set(this, new Record(Record.BOTTOM));
+            this.allLeaks = allLeaks;
+        }
+
+        /**
+         * Adds an access record that includes the current stack trace to the leak.
+         */
+        public void record() {
+            Record oldHead;
+            Record newHead;
+            boolean dropped;
+            do {
+                Record prevHead;
+                if ((prevHead = oldHead = headUpdater.get(this)) == null) {
+                    // already closed.
+                    return;
+                }
+                final int numElements = oldHead.pos + 1;
+                if (numElements >= TARGET_RECORDS) {
+                    final int backOffFactor = Math.min(numElements - TARGET_RECORDS, 30);
+                    if (dropped = Randomness.get().nextInt(1 << backOffFactor) != 0) {
+                        prevHead = oldHead.next;
+                    }
+                } else {
+                    dropped = false;
+                }
+                newHead = new Record(prevHead);
+            } while (headUpdater.compareAndSet(this, oldHead, newHead) == false);
+            if (dropped) {
+                droppedRecordsUpdater.incrementAndGet(this);
+            }
+        }
+
+        private boolean dispose() {
+            clear();
+            return allLeaks.remove(this);
+        }
+
+        /**
+         * Stop tracking the object that this leak was created for.
+         *
+         * @param trackedObject the object that this leak was originally created for
+         * @return true if the leak was released by this call, false if the leak had already been released
+         */
+        public boolean close(T trackedObject) {
+            assert trackedHash == System.identityHashCode(trackedObject);
+            try {
+                if (allLeaks.remove(this)) {
+                    // Call clear so the reference is not even enqueued.
+                    clear();
+                    headUpdater.set(this, null);
+                    return true;
+                }
+                return false;
+            } finally {
+                reachabilityFence0(trackedObject);
+            }
+        }
+
+        private static void reachabilityFence0(Object ref) {
+            if (ref != null) {
+                synchronized (ref) {
+                    // empty on purpose
+                }
+            }
+        }
+
+        @Override
+        public String toString() {
+            Record oldHead = headUpdater.getAndSet(this, null);
```

Comment:
Pretty yucky that we only get to call this once, can we tidy that up at all?

Reply:
I think we could without breaking anything just invoke `get()` here. No need for this optimization since unlike Netty we're not using this in production. Will adjust accordingly, that has the nice side effect of allowing some IDE debugging on these objects :)